### ISSUE TITLE
FP: "is_array(json_decode(something))" _does_ make sense

### ISSUE THREAD

**Issue by tlamy** [kalessil/phpinspectionsea#1539]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Plugin         | Php Inspections (EA Ultimate) 3.0.8.1 |
| Language level | PHP 7.1                        |

Current behaviour
--
![image](https://user-images.githubusercontent.com/1841537/78141915-76e00f80-742c-11ea-84fd-47213572ae44.png)

Expected behaviour
--
Inspection should not warn, see https://3v4l.org/8oqaE

Environment details
---
PhpStorm 2020.1 Beta
Build #PS-201.6668.30, built on March 26, 2020
PhpStorm EAP User
Expiration date: April 25, 2020
Runtime version: 11.0.6+8-b765.17 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.14.5
GC: G1 Young Generation, G1 Old Generation
Memory: 2048M
Cores: 6
Registry: run.processes.with.pty=TRUE
Non-Bundled Plugins: zielu.gittoolbox, String Manipulation, com.alayouni.ansiHighlight, com.intellij.ideolog, ru.adelf.idea.dotenv, NEON support, com.kalessil.phpStorm.phpInspectionsUltimate, de.espend.idea.php.annotation, org.psliwa.idea.composer, izhangzhihao.rainbow.brackets, org.zalando.intellij.swagger

**Comment by riconeitzel** [comment]
can second that:

```
        $archiveData = json_decode($archiveData, true);

        if ($archiveData === null) {
            return true;
        }

```

throws `[EA] Makes no sense, because this type is not defined in annotations` at me.

**Comment by kalessil** [comment]
Thank you for reporting @tlamy: I'll look into this.

**Comment by kalessil** [comment]
@tlamy, @riconeitzel: checked, seems the $filter/$archiveData is also method/parameter with declared type missing array/null. While the code is perfectly fine for PHP itself, the inspection correctly reports deviation from declared types.

So I think we need a new setting here to disable this specific behavior. Will this work for you, or perhaps there are better ideas?

**Comment by tlamy** [comment]
The only thing one can be sure of is that json_decode(anything, true) can never return stdClass, if the second flag is false the return type is int|float|string|bool|object|array|null - did i forget any type?

**Comment by riconeitzel** [comment]
@kalessil ich changed my code when I noticed, phpea takes it from the signature. 

Perhaps it would be enough to just improve the "error message". Something like "you reassigned the variable but it doesn't match the methods parameter signature. this is considered bad style …" something like this.

**Comment by kalessil** [comment]
@tlamy: nope, types seems to be complete. The problem is IDE types resolver - sometimes it doesn't merge/expand types correctly. Like in your example. The only working solution I came so far is to not reuse parameters and variables for those cases - then for each variable types solving gets significantly better.

**Comment by kalessil** [comment]
Okay, I update the messages based on the discussion here.
